pkg/util/limiter/memory_limiter.go

```diff
@@ -0,0 +1,379 @@
+	f.DurationVar(&cfg.MaxWaitDuration, prefix+"memory-limiter.max-wait-duration", 5*time.Second, "Configures the maximum wait duration. If a goroutine were blocked for more than the max-wait-duration, it'll continue execution even under memory pressure.")
+}
+
+// MemoryLimiter provides a way for the service to detect memory pressure and slow itself down.
+type MemoryLimiter interface {
+	// NewJob creates a new job and adds to the job queue.
+	NewJob() Job
+
+	// Wait checks if the state of the system and blocks () until the desired state is achieved or context is done.
+	// If allowFirst is true, the first request in the queue is always allowed to succeed. This can be used to prevent deadlocks.
+	Wait(ctx context.Context, id JobID, desiredState State, allowFirst bool, failFast bool) error
+
+	// enqueue adds a job to the queue.
+	enqueue(j Job)
+
+	// dequeue removes a job from the queue.
+	dequeue(j Job)
+}
+
+type JobID int64
+
+// Job is a piece of work that the service is assigned to do.
+type Job interface {
+	// ID returns the id of the job
+	ID() JobID
+
+	// Complete ends the job and removes it from the queue.
+	Complete()
+
+	// Continue waits until the state is yellow or green.
+	Continue(ctx context.Context) error
+}
+
+type job struct {
+	failFast bool
+	id       JobID
+	limiter  MemoryLimiter
+}
+
+func (j *job) ID() JobID {
+	return j.id
+}
+
+func (j *job) Complete() {
+	j.limiter.dequeue(j)
+}
+
+func (j *job) Continue(ctx context.Context) error {
+	return j.limiter.Wait(ctx, j.id, Yellow, true, j.failFast)
+}
+
+// AddJobToContext is the helper function to add the job to context.
+func AddJobToContext(ctx context.Context, job Job) context.Context {
+	return context.WithValue(ctx, mlJobCtxkey, job)
+}
+
+// JobFromContext is the helper function to get the job from context.
+func JobFromContext(ctx context.Context) Job {
+	j, ok := ctx.Value(mlJobCtxkey).(Job)
+	if !ok {
+		return &job{
+			id:      0,
+			limiter: &NoOpMemoryLimiter{},
+		}
+	}
+	return j
+}
+
+type jobCtxKey struct{}
+
+var (
+	mlJobCtxkey = &jobCtxKey{}
+)
+
+// NewMemoryLimiter returns a new memorylimiter.
+func NewMemoryLimiter(cfg *MemoryLimiterConfig, r prometheus.Registerer, logger log.Logger) (MemoryLimiter, error) {
+	if !cfg.Enabled {
+		return &NoOpMemoryLimiter{}, nil
+	}
+
+	if cfg.CheckInterval <= 0 {
+		return nil, errCheckIntervalOutOfRange
+	}
+	if cfg.RedThresholdBytes == 0 || cfg.YellowThresholdBytes == 0 {
+		return nil, errThresholdOutOfRange
+	}
+	if cfg.YellowThresholdBytes >= cfg.RedThresholdBytes {
+		return nil, errInvlaidThreshold
+	}
+
+	ml := &HeapMemoryLimiter{
+		failFast:         cfg.FailFast,
+		memCheckInterval: cfg.CheckInterval,
+		redThreshold:     cfg.RedThresholdBytes,
+		yellowThreshold:  cfg.YellowThresholdBytes,
+		ticker:           time.NewTicker(cfg.CheckInterval),
+		readMemStatsFn:   runtime.ReadMemStats,
```

Comment:
I still a little concerned about the `stopTheWorld` when calling `runtime.ReadMemStats`.

Did you see any problem with this? I guess this would be a bigger problem on the ingestion path but maybe here should be ok?

@alvinlin123 WDYT?